[Security Solution][Detections] Related Integrations & Required Fields Feedback & Fixes

spong opened the pull request:
## Summary

Addressing the following feedback from https://github.com/elastic/kibana/pull/132847:

- [X] On the Rule Management page package name is used instead of package title when the package has only 1 integration:
- [X] move integrations_popover to `related_integrations` directory
- [X] update useInstalledIntegrations to always return array of installedIntegration
- [X] move useInstalledIntegrations to `related_integrations` directory
- [X] Slight update to copy in Rules Table popover
- [X] Ok to use Rule Details UI within Rules Table popover content
- [X] Sort integrations alphabetically
- [X] Update left padding on version mis-match tooltip
- [X] https://github.com/elastic/kibana/issues/133291
- [X] https://github.com/elastic/kibana/issues/133269
- [X] https://github.com/elastic/kibana/issues/133600
- [X]  Add Kibana Advanced Setting for disabling integrations badge on Rules Table
<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/171750790-ffa2e3ef-dd7a-499c-9b08-89bafc06dd50.png" />
</p>

- [ ]  Adds tests
  - [x] `useInstalledIntegrations` hook 
  - [X] relatedIntegrations utils
  - [x] IntegrationDescription
- [ ]  Add loaders where necessary since there can now be API delay
  - May hold off on loaders as transition from no installed integrations -> installed integrations isn't too bad as-is

##### Updated integrations popover content on Rules Table to match Rule Details design
<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/172263941-3e948b41-7ef7-4281-8354-57e77ddeb433.png" />
</p>


In discussions with @banderror reviewing the different integration states (uninstalled, installed, enabled, and agents enrolled), we are now capturing the distinction between `Installed` and `Enabled` so that we don't confuse users when a package is installed but the integration isn't enabled/configured. I also added tooltips for clarifying each state and what action the user should perform to ensure compatibility. In collab with @yiyangliu9286 @jethr0null (comments below) -- we've consolidated to a single `Installed: enabled` badge, and updated `Uninstalled` to `Not installed` as well.


##### Tooltips

<details><summary>Not installed</summary>
<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/172264210-00064485-2df9-408e-953b-9294f16dedf2.png" />
</p>
</details>





<details><summary>Installed</summary>
<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/172263672-67b641cd-5895-464a-8897-f26bd0a61073.png" />
</p>
</details>




<details><summary>Installed: enabled</summary>
<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/172263783-563ea48d-c96c-4519-87b4-7076582f5da2.png" />
</p>
</details>



### Checklist

Delete any items that are not applicable to this PR.

- [X] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [X] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
  - Collaborating with docs teams on this dedicated docs issue: https://github.com/elastic/security-docs/issues/2015
- [X] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [X] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

spong commented on packages/kbn-field-types/src/kbn_field_types_factory.ts:
```diff
@@ -21,6 +21,7 @@ export const createKbnFieldTypes = (): KbnFieldType[] => [
     esTypes: [
       ES_FIELD_TYPES.STRING,
       ES_FIELD_TYPES.TEXT,
+      ES_FIELD_TYPES.MATCH_ONLY_TEXT,
```
@elastic/kibana-app-services -- in leveraging the `castEsToKbnFieldTypeName` utility I found this missing [ES type](https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html#match-only-text-field-type). I imagine this is fine as it's a sibling type to `text`, but was curious here since `sortable: true,` is set above even though it includes `ES_FIELD_TYPES.TEXT` (which is not sortable) as well?

spong replied:
Also, I had [previously updated](https://github.com/elastic/kibana/pull/131475/files#diff-d79a8297783f3177da25dd13fe807425d9136a0e235fe170f7c0a61f2448dacaR23) `FieldIcon` to support `match_only_text` as well, but this was before I noticed the distinction between `ES_FIELD_TYPES` and `KBN_FIELD_TYPES` with the `FieldIcon` component. Seems the `FieldIcon` component kind of supports both (with at least `keyword` coming from ES_FIELD_TYPES) -- should I undo my previous change in support of `match_only_text` there as well, or it fine to leave it as-is?

spong replied:
Moved to dedicated PR -- please see https://github.com/elastic/kibana/pull/133690

yiyangliu9286 commented:
I agree with the badge colour usage here to be consistent with what we use in ML Job enabled/disabled UI. On the different statuses itself though I wonder if it makes sense to combine the `Installed` and `Enabled` as `Installed: enabled` rather than two badges since the Tooltip content is similar. So we have 3 different status in total.

1. `Uninstalled` Show tooltip:
<img width="700" alt="171966180-8a2be278-d135-4770-8326-729d23e0bd7f" src="https://user-images.githubusercontent.com/81987435/172236227-615cb502-cbc0-4862-b643-91b3826ae6b1.png">

2. `Installed` Show tooltip:
<img width="697" alt="171966237-d47aee81-a9d9-4e83-b7ad-904989444900" src="https://user-images.githubusercontent.com/81987435/172236319-12250d1d-e6ed-45a1-8a9a-5585b61c14a2.png">

3. `Installed: enabled` Show tooltip
`Integration is installed. An integration policy with the required configuration exists. Ensure Elastic Agents are assigned this policy to ingest compatible events.`

jethr0null commented:
@spong I’m on board with @yiyang’s suggestions as well. My only other thought/question is whether it makes sense to use `Available` instead of `Uninstalled` to avoid any ambiguity (e.g. uninstalled being interpreted as “was once installed and has since been removed”. I don't feel strongly about changing this if everyone else agrees that `uninstalled` makes more sense.

spong commented:
Thanks for the feedback @yiyangliu9286 and @jethr0null! 🙂 

I've combined the `Installed` and `Enabled` badges, updated `Uninstalled` to `Not installed` (from our other discussion), and updated the badge colors such that `Installed` is the green `success`, and `Not installed` is the muted gray. I've updated the images in the description, so please check those out to confirm these changes 👍 .

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/use_installed_integrations.tsx:
```diff
@@ -18,23 +18,25 @@ export interface UseInstalledIntegrationsArgs {
       onError: (e) => {
-        addError(e, { title: i18n.INSTALLED_INTEGRATIONS_FETCH_FAILURE });
+        // TODO: Suppress if always failing in cases where fleet isn't setup, etc
+        addError(e, { title: i18n.INTEGRATIONS_FETCH_FAILURE });
```
I'm thinking we should just suppress this error toast for now until we have a chance to test all the RBAC scenarios on when/where it can fail that we've outlined in the test plan. Would rather fail silently here than spam certain deployments that have limited privileges, access to fleet, etc...

banderror replied:
@spong totally agree!

banderror commented:
While testing locally, found a few bugs. Will post these right away and continue testing and reviewing the code.

## Version mismatch icon's color is off

When the dark mode is off:

<img width="918" alt="Screenshot 2022-06-08 at 21 34 24" src="https://user-images.githubusercontent.com/7359339/172703828-077b685e-b9b1-4412-a00e-e9d217fac115.png">

## Links include a semver requirement instead of a target version

Looks like it happens for installed integrations, and works properly for those that are not installed. Both pages are affected.

<img width="958" alt="Screenshot 2022-06-08 at 21 37 28" src="https://user-images.githubusercontent.com/7359339/172704107-5e93d137-ea82-44b8-917c-9aa48a8e118c.png">

<img width="988" alt="Screenshot 2022-06-08 at 21 37 59" src="https://user-images.githubusercontent.com/7359339/172704398-a19ff714-81e6-4f81-b27f-b10a6d286004.png">

ndjson for repro:

```
{"id":"6cc39c80-da3a-11ec-9fce-65c1a0bee900","updated_at":"2022-05-23T01:48:23.422Z","updated_by":"elastic","created_at":"2022-05-23T01:48:20.940Z","created_by":"elastic","name":"Rule with Related Integrations, Required Fields, and Setup Guide","tags":["Test"],"interval":"5m","enabled":false,"description":"See Related Integrations, Required Fields, and Setup Guide on the Rule Details and Rule Management pages.","risk_score":47,"severity":"medium","license":"Elastic License v2","output_index":".siem-signals-default","meta":{"from":"5m"},"rule_name_override":"message","timestamp_override":"event.ingested","author":["Elastic"],"false_positives":[],"from":"now-600s","rule_id":"2c66bf23-6ae9-4eb2-859e-446bea181ae0","max_signals":10000,"risk_score_mapping":[],"severity_mapping":[],"threat":[],"to":"now","references":[],"version":7,"exceptions_list":[],"immutable":false,"related_integrations":[{"package":"system","version":"~1.6.4"},{"package":"aws","integration":"cloudfront","version":"~1.11.0"},{"package":"aws","integration":"cloudtrail","version":"~1.11.0"},{"package":"aws","integration":"route53","version":"~1.11.0"},{"package":"unknown","integration":"unknown","version":"^1.2.3"},{"package":"unknown2","version":"4.5.6"}],"required_fields":[{"ecs":true,"name":"event.code","type":"keyword"},{"ecs":true,"name":"message","type":"match_only_text"},{"ecs":false,"name":"winlog.event_data.AttributeLDAPDisplayName","type":"keyword"},{"ecs":false,"name":"winlog.event_data.AttributeValue","type":"keyword"},{"ecs":false,"name":"winlog.event_data.ShareName","type":"keyword"},{"ecs":false,"name":"winlog.event_data.RelativeTargetName","type":"keyword"},{"ecs":false,"name":"winlog.event_data.AccessList","type":"keyword"}],"setup":"## Config\n\nThe 'Audit Detailed File Share' audit policy must be configured (Success Failure).\nSteps to implement the logging policy with with Advanced Audit Configuration:\n\n```\nComputer Configuration > \nPolicies > \nWindows Settings > \nSecurity Settings > \nAdvanced Audit Policies Configuration > \nAudit Policies > \nObject Access > \nAudit Detailed File Share (Success,Failure)\n```\n\nThe 'Audit Directory Service Changes' audit policy must be configured (Success Failure).\nSteps to implement the logging policy with with Advanced Audit Configuration:\n\n```\nComputer Configuration > \nPolicies > \nWindows Settings > \nSecurity Settings > \nAdvanced Audit Policies Configuration > \nAudit Policies > \nDS Access > \nAudit Directory Service Changes (Success,Failure)\n```\n","type":"query","language":"kuery","index":["logs-*"],"query":"host.name:*","filters":[],"throttle":"no_actions","actions":[]}
{"id":"6cc39c80-da3a-11ec-9fce-65c1a0bee901","updated_at":"2022-05-23T01:48:23.422Z","updated_by":"elastic","created_at":"2022-05-23T01:48:20.940Z","created_by":"elastic","name":"Rule with Related Integrations","tags":["Test"],"interval":"5m","enabled":false,"description":"See Related Integrations on the Rule Details and Rule Management pages.","risk_score":47,"severity":"medium","license":"Elastic License v2","output_index":".siem-signals-default","meta":{"from":"5m"},"rule_name_override":"message","timestamp_override":"event.ingested","author":["Elastic"],"false_positives":[],"from":"now-600s","rule_id":"2c66bf23-6ae9-4eb2-859e-446bea181ae1","max_signals":10000,"risk_score_mapping":[],"severity_mapping":[],"threat":[],"to":"now","references":[],"version":7,"exceptions_list":[],"immutable":false,"related_integrations":[{"package":"unknown","version":"4.5.6"},{"package":"unknown","integration":"unknown","version":"^1.2.3"},{"package":"system","version":"~1.6.4"},{"package":"aws","integration":"cloudfront","version":"~1.11.0"},{"package":"aws","integration":"cloudtrail","version":"~1.11.0"},{"package":"aws","integration":"route53","version":"~123.456.789"},{"package":"azure","integration":"activitylogs","version":"~99.99.99"},{"package":"azure","integration":"platformlogs","version":"100.x.x"}],"required_fields":[],"setup":"","type":"query","language":"kuery","index":["logs-*"],"query":"host.name:*","filters":[],"throttle":"no_actions","actions":[]}
{"id":"6cc39c80-da3a-11ec-9fce-65c1a0bee902","updated_at":"2022-05-23T01:48:23.422Z","updated_by":"elastic","created_at":"2022-05-23T01:48:20.940Z","created_by":"elastic","name":"Rule with Required Fields","tags":["Test"],"interval":"5m","enabled":false,"description":"See Required Fields on the Rule Details page.","risk_score":47,"severity":"medium","license":"Elastic License v2","output_index":".siem-signals-default","meta":{"from":"5m"},"rule_name_override":"message","timestamp_override":"event.ingested","author":["Elastic"],"false_positives":[],"from":"now-600s","rule_id":"2c66bf23-6ae9-4eb2-859e-446bea181ae2","max_signals":10000,"risk_score_mapping":[],"severity_mapping":[],"threat":[],"to":"now","references":[],"version":7,"exceptions_list":[],"immutable":false,"related_integrations":[],"required_fields":[{"ecs":true,"name":"event.code","type":"keyword"},{"ecs":true,"name":"message","type":"match_only_text"},{"ecs":false,"name":"winlog.event_data.AttributeLDAPDisplayName","type":"keyword"},{"ecs":false,"name":"winlog.event_data.AttributeValue","type":"keyword"},{"ecs":false,"name":"winlog.event_data.ShareName","type":"keyword"},{"ecs":false,"name":"winlog.event_data.RelativeTargetName","type":"keyword"},{"ecs":false,"name":"winlog.event_data.AccessList","type":"keyword"}],"setup":"","type":"query","language":"kuery","index":["logs-*"],"query":"host.name:*","filters":[],"throttle":"no_actions","actions":[]}
{"id":"6cc39c80-da3a-11ec-9fce-65c1a0bee903","updated_at":"2022-05-23T01:48:23.422Z","updated_by":"elastic","created_at":"2022-05-23T01:48:20.940Z","created_by":"elastic","name":"Rule with Setup Guide","tags":["Test"],"interval":"5m","enabled":false,"description":"See Setup Guide on the Rule Details page.","risk_score":47,"severity":"medium","license":"Elastic License v2","output_index":".siem-signals-default","meta":{"from":"5m"},"rule_name_override":"message","timestamp_override":"event.ingested","author":["Elastic"],"false_positives":[],"from":"now-600s","rule_id":"2c66bf23-6ae9-4eb2-859e-446bea181ae3","max_signals":10000,"risk_score_mapping":[],"severity_mapping":[],"threat":[],"to":"now","references":[],"version":7,"exceptions_list":[],"immutable":false,"related_integrations":[],"required_fields":[],"setup":"## Config\n\nThe 'Audit Detailed File Share' audit policy must be configured (Success Failure).\nSteps to implement the logging policy with with Advanced Audit Configuration:\n\n```\nComputer Configuration > \nPolicies > \nWindows Settings > \nSecurity Settings > \nAdvanced Audit Policies Configuration > \nAudit Policies > \nObject Access > \nAudit Detailed File Share (Success,Failure)\n```\n\nThe 'Audit Directory Service Changes' audit policy must be configured (Success Failure).\nSteps to implement the logging policy with with Advanced Audit Configuration:\n\n```\nComputer Configuration > \nPolicies > \nWindows Settings > \nSecurity Settings > \nAdvanced Audit Policies Configuration > \nAudit Policies > \nDS Access > \nAudit Directory Service Changes (Success,Failure)\n```\n","type":"query","language":"kuery","index":["logs-*"],"query":"host.name:*","filters":[],"throttle":"no_actions","actions":[]}
{"id":"6cc39c80-da3a-11ec-9fce-65c1a0bee904","updated_at":"2022-05-23T01:48:23.422Z","updated_by":"elastic","created_at":"2022-05-23T01:48:20.940Z","created_by":"elastic","name":"Rule without the new fields","tags":["Test"],"interval":"5m","enabled":false,"description":"See no Related Integrations, Required Fields, and Setup Guide on the Rule Details and Rule Management pages.","risk_score":47,"severity":"medium","license":"Elastic License v2","output_index":".siem-signals-default","meta":{"from":"5m"},"rule_name_override":"message","timestamp_override":"event.ingested","author":["Elastic"],"false_positives":[],"from":"now-600s","rule_id":"2c66bf23-6ae9-4eb2-859e-446bea181ae4","max_signals":10000,"risk_score_mapping":[],"severity_mapping":[],"threat":[],"to":"now","references":[],"version":7,"exceptions_list":[],"immutable":false,"type":"query","language":"kuery","index":["logs-*"],"query":"host.name:*","filters":[],"throttle":"no_actions","actions":[]}
{"exported_count":5,"exported_rules_count":5,"missing_rules":[],"missing_rules_count":0,"exported_exception_list_count":0,"exported_exception_list_item_count":0,"missing_exception_list_item_count":0,"missing_exception_list_items":[],"missing_exception_lists":[],"missing_exception_lists_count":0}
```

spong commented:
> While testing locally, found a few bugs. Will post these right away and continue testing and reviewing the code.
> 
> ### Version mismatch icon's color is not readable
> When the dark mode is off:
> 
> ### Links include a semver requirement instead of a target version
> Looks like it happens ~for installed integrations, and works properly for those that are not installed~ when there's a version mismatch. Both pages are affected.


Thanks @banderror! Should be resolved as of [b0ba541](https://github.com/elastic/kibana/pull/133050/commits/b0ba541b75057af005f170d6b0b32945d2ff4984). I've updated the icon color to be `warning` as it is in the fleet UI, ensure the left-margin was being applied (it was set, but was being overriden by the `EuiIconTip`), and lastly fixed the version calculation and started adding our tests from the test plan. 

There might still be another bug in here with the version calculation, so will finish adding the remaining tests and push those next 👍 

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/172731049-f7b1ae06-2b1c-4227-9bde-891446bf6bba.png" />
</p>

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/integrations_popover/index.tsx:
```diff
@@ -0,0 +1,116 @@
+    allInstalledIntegrations
+  );
+
+  const totalRelatedIntegrationsInstalled = integrationDetails.filter((i) => i.is_enabled).length;
```
Nit: `totalRelatedIntegrationsEnabled`

spong replied:
Ended up being resolved in https://github.com/elastic/kibana/pull/134299

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/integrations_popover/index.tsx:
```diff
@@ -0,0 +1,116 @@
+    <IntegrationsPopoverWrapper
+      alignItems="center"
+      gutterSize="s"
+      data-test-subj={'IntegrationsWrapper'}
```
Nit: `data-test-subj="IntegrationsWrapper"`
Same for some of the props below.

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/mock.ts:
```diff
@@ -0,0 +1,100 @@
+  RelatedIntegrationArray,
+} from '../../../../../common/detection_engine/schemas/common';
+
+export const relatedIntegrations: RelatedIntegrationArray = [
+  {
+    package: 'system',
+    version: '1.6.4',
+  },
+  {
+    package: 'aws',
+    integration: 'cloudtrail',
+    version: '~1.11.0',
+  },
+];
```
I think we should keep all this mock data near each type (like `RelatedIntegrationArray`) definition. I can move it there in a follow-up PR. I should have added some mock data in https://github.com/elastic/kibana/pull/132409

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/mock.ts:
```diff
@@ -0,0 +1,100 @@
+  },
+];
+
+export const integrationDetailsUninstalled: IntegrationDetails = {
+  package_name: 'test',
+  package_title: 'Test',
+  package_version: '1.2.3',
+  integration_name: 'integration',
+  integration_title: 'Integration',
+  is_enabled: false,
+  is_installed: false,
+  target_version: '1.2.3',
+  version_satisfied: false,
+};
+
+export const integrationDetailsInstalled: IntegrationDetails = {
+  package_name: 'test',
+  package_title: 'Test',
+  package_version: '1.2.3',
+  integration_name: 'integration',
+  integration_title: 'Integration',
+  is_enabled: false,
+  is_installed: true,
+  target_version: '1.2.3',
+  version_satisfied: true,
+};
+
+export const integrationDetailsEnabled: IntegrationDetails = {
+  package_name: 'test',
+  package_title: 'Test',
+  package_version: '1.1.3',
+  integration_name: 'integration',
+  integration_title: 'Integration',
+  is_enabled: true,
+  is_installed: true,
+  target_version: '1.3.3',
+  version_satisfied: false,
+};
```
Can we have an `integration_details.ts` file and then move this mock data to `integration_details.mock.ts`?

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/translations.ts:
```diff
@@ -0,0 +1,94 @@
+    {
+      values: { installedVersion, targetVersion },
+      defaultMessage:
+        'Version mismatch -- please resolve! Installed version `{installedVersion}` when target version `{targetVersion}`',
```
We define `targetVersion` as the version that we redirect the user to, I believe. Do you think we should display it here in this tooltip, or rather the version requirement (`relatedIntegration.version`)?

spong replied:
As discussed, this is a naming issue we need to clean up between `target_version` and the concrete version we get from that. Will clean up the naming as part of storing all data on the `RelatedIntegrationsDetails` object.

spong replied:
Ended up being resolved in https://github.com/elastic/kibana/pull/134299

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/utils.tsx:
```diff
@@ -0,0 +1,113 @@
+/*
```
Not a big fan of `utils.ts`, `types.ts`, and stuff like that. Could we split this into separate files?
1. One for `getIntegrationLink` (I'd make it a react FC called `RelatedIntegrationLink`)
2. And the other for `IntegrationDetails` and `getInstalledRelatedIntegrations` (`integration_details.ts`) - that one would actually contain our domain logic, so not "utils" please! 🙂

spong replied:
Ended up being resolved in https://github.com/elastic/kibana/pull/134299

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/utils.tsx:
```diff
@@ -0,0 +1,113 @@
+    ? integration.package_version
+    : integration.target_version;
+
+  const integrationURL =
+    version !== ''
+      ? `${basePath}/app/integrations/detail/${packageName}-${version}/overview${
+          integrationName ? `?integration=${integrationName}` : ''
+        }`
+      : `${basePath}/app/integrations/detail/${packageName}`;
```
I think this calculation should be inside `getInstalledRelatedIntegrations` and we should expose a `target_url` from `IntegrationDetails`. It's a piece of domain logic and it'd be much better to keep it separate from rendering.

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/utils.tsx:
```diff
@@ -0,0 +1,113 @@
+ * @param integrations
+ * @param installedIntegrations
+ */
+export const getInstalledRelatedIntegrations = (
```
Nit: `calculateIntegrationDetails`

spong replied:
Ended up being resolved in https://github.com/elastic/kibana/pull/134299

banderror reviewed (commented):
Almost there! Reviewed most of the changes and have some suggestions, so please check the comments I left @spong.

Still have to carefully scan the `getInstalledRelatedIntegrations` logic and review the test coverage. Will do it tomorrow morning.

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/utils.tsx:
```diff
@@ -0,0 +1,113 @@
+    );
+
+    if (match != null) {
+      // Version check e.g. fleet match `1.2.3` satisfies rule dependency `~1.2.1`
+      const versionSatisfied = semver.satisfies(match.package_version, i.version);
+      const packageVersion = versionSatisfied
+        ? i.version
+        : semver.valid(semver.coerce(i.version)) ?? '';
```
I'm thinking - if the version requirement is satisfied, shouldn't we use the installed version? Here in the code, it would be `match.package_version`:

```ts
      const targetVersion = versionSatisfied
        ? match.package_version
        : semver.valid(semver.coerce(i.version)) ?? '';
```

`i.version` here is the related integration's `version` which is `~1.2.3` etc.

So to me, it looks like a bug, but it works as expected locally and there are some unit tests that seem to cover this code path, so I'm not sure about my sanity 😂

banderror commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/utils.tsx:
```diff
@@ -0,0 +1,113 @@
+    }
+  });
+
+  return integrationDetails.sort((a, b) => {
+    if (a.integration_title != null && b.integration_title != null) {
+      return a.integration_title.localeCompare(b.integration_title);
+    } else if (a.integration_title != null) {
+      return a.integration_title.localeCompare(b.package_title);
+    } else if (b.integration_title != null) {
+      return a.package_title.localeCompare(b.integration_title);
+    } else {
+      return a.package_title.localeCompare(b.package_title);
+    }
+  });
```
Nit: Could this do the same?

```ts
  return integrationDetails.sort((a, b) => {
    const aTitle = a.integration_title ?? a.package_title;
    const bTitle = b.integration_title ?? b.package_title;
    return aTitle.localeCompare(bTitle);
  });
```

Also, if we always set `integration_title` in `IntegrationDetails`, could be even simpler:

```ts
  return integrationDetails.sort((a, b) => {
    return a.integration_title.localeCompare(b.integration_title);
  });
```

Then, in `getIntegrationLink` we wouldn't need to do ?? in `const integrationTitle = integration.integration_title ?? integration.package_title;`

spong replied:
Yeah I think we can always set `integrationTitle` to clean this up and then keep `integrationName` `undefined` if it's a single integration package (i.e. `system` package).

banderror reviewed (commented):
Last comments! Finished reviewing the code changes and tested it once again.

Couldn't find any new bugs, and the two bugs I reported above are fixed! 🙌

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/integrations_description/index.test.tsx:
```diff
@@ -0,0 +1,37 @@
+      allInstalledIntegrations
+    );
+    render(<IntegrationDescription integration={integrationDetails[0]} />);
+    expect(screen.getByTestId('integrationLink')).toHaveTextContent('Aws Cloudtrail');
```
Nice catch -- will rename!

spong replied:
Ended up being resolved in https://github.com/elastic/kibana/pull/134299

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/integrations_description/index.tsx:
```diff
@@ -0,0 +1,109 @@
+  );
+};
+
+export const buildRelatedIntegrationsDescription = (
+  label: string,
+  relatedIntegrations: RelatedIntegrationArray | undefined
+): ListItems[] => {
+  if (relatedIntegrations == null || relatedIntegrations.length === 0) {
+    return [];
+  }
+
+  return [
+    {
+      title: label,
+      description: <RelatedIntegrationsDescription integrations={relatedIntegrations} />,
+    },
+  ];
+};
```
Will move this specific function back to `rules/description_step/helpers.tsx` 👍

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/integrations_popover/index.tsx:
```diff
@@ -0,0 +1,116 @@
+const MemoizedIntegrationsPopover = React.memo(IntegrationsPopoverComponent);
+MemoizedIntegrationsPopover.displayName = 'IntegrationsPopover';
+
+export const IntegrationsPopover =
+  MemoizedIntegrationsPopover as typeof IntegrationsPopoverComponent;
```
This was leftover from the base component I used for making this -- will clean up

spong commented on x-pack/plugins/security_solution/server/ui_settings.ts:
```diff
@@ -250,6 +251,22 @@ export const initUiSettings = (
+      ),
+      type: 'boolean',
+      category: [APP_ID],
+      requiresPageReload: true,
```
Yeah we can change to `false` -- I had tested in multiple tabs and noticed the opposite behavior so had set as `true`, but this appears to be a cache issue.

banderror reviewed (approved):
Synced on the comments with @spong and decided to address them in a follow-up PR.
🚀 🚀 🚀

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.3|**Backport failed because of merge conflicts**|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 133050
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
<!--- Backport version: 7.4.0 -->

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/integrations_description/index.test.tsx:
```diff
@@ -0,0 +1,37 @@
+
+describe('IntegrationDescription', () => {
+  test('Shows total events returned', () => {
+    const { data: allInstalledIntegrations } = useInstalledIntegrations({ packages: [] });
```
Ended up being resolved in https://github.com/elastic/kibana/pull/134299

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/integrations_description/index.test.tsx:
```diff
@@ -0,0 +1,37 @@
+  test('Shows total events returned', () => {
+    const { data: allInstalledIntegrations } = useInstalledIntegrations({ packages: [] });
+
+    const integrationDetails = getInstalledRelatedIntegrations(
+      relatedIntegrations,
+      allInstalledIntegrations
+    );
```
Ended up being resolved in https://github.com/elastic/kibana/pull/134299

spong commented on x-pack/plugins/security_solution/public/detections/components/rules/related_integrations/integrations_description/index.tsx:
```diff
@@ -0,0 +1,109 @@
+        <PaddedBadge color={badgeColor}>{badgeText}</PaddedBadge>
+      </EuiToolTip>
+      {integration.is_installed && !integration.version_satisfied && (
+        <VersionWarningIconContainer>
+          <EuiIconTip
+            type={'alert'}
+            color={'warning'}
+            content={i18n.INTEGRATIONS_INSTALLED_VERSION_TOOLTIP(
+              integration.package_version,
+              integration.target_version
+            )}
+          />
+        </VersionWarningIconContainer>
```
Ended up being resolved in https://github.com/elastic/kibana/pull/134299
